### PR TITLE
Add #nullable for Endpoint related classes

### PR DESCRIPTION
**Pull request by terencefan**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 1]
```diff
@@ -125,12 +123,12 @@ public ServiceEndpoint(string connectionString, EndpointType type = EndpointType
         var result = ConnectionStringParser.Parse(connectionString);
         AccessKey = result.AccessKey;
         _serviceEndpoint = result.Endpoint;
-        ClientEndpoint = result.ClientEndpoint;
-        ServerEndpoint = result.ServerEndpoint;
         EndpointType = type;
         Name = name;
 
         Endpoint = BuildEndpointString(_serviceEndpoint);
+        ClientEndpoint = result.ClientEndpoint ?? _serviceEndpoint;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs]: keep it as null if result.ClientEndpoint is null?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 2]
```diff
@@ -125,12 +123,12 @@ public ServiceEndpoint(string connectionString, EndpointType type = EndpointType
...
+        ServerEndpoint = result.ServerEndpoint ?? _serviceEndpoint;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs]: keep it as null if ServerEndpoint is null?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 3]
```diff
@@ -172,8 +169,8 @@ public ServiceEndpoint(Uri endpoint,
 
         AudienceBaseUrl = BuildAudienceBaseUrlEndWithSlash(_serviceEndpoint);
         Endpoint = BuildEndpointString(_serviceEndpoint);
-        ServerEndpoint = serverEndpoint;
-        ClientEndpoint = clientEndpoint;
+        ServerEndpoint = serverEndpoint ?? endpoint;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs]: same here, no need to fallback

File: src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs [hunk 4]
```diff
@@ -155,13 +159,13 @@ internal static ParsedConnectionString Parse(string connectionString)
         };
     }
 
-    internal static bool TryGetEndpointUri(string endpoint, out Uri uriResult)
+    private static bool TryCreateEndpointUri(string endpoint, out Uri? uriResult)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs:161]: could be NotNullWhenTrue

File: src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs [hunk 5]
```diff
@@ -1,1533 +1,1532 @@
...
-    /// <summary>
-    /// Implements the Azure SignalR Service Protocol.
-    /// </summary>
-    public class ServiceProtocol : IServiceProtocol
-    {
-        private static readonly IDictionary<string, ReadOnlyMemory<byte>> EmptyReadOnlyMemoryDictionary = new Dictionary<string, ReadOnlyMemory<byte>>();
+namespace Microsoft.Azure.SignalR.Protocol;
 
-        private static readonly IDictionary<string, StringValues> EmptyStringValuesDictionaryIgnoreCase = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
+#nullable enable
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs:15]: could remove, the whole project is nullable now

File: src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs (hunk 4 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs:161]: Not possible, netstandard 2.0 does not support this grammar.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 6]
```diff
@@ -125,12 +123,12 @@ public ServiceEndpoint(string connectionString, EndpointType type = EndpointType
...
         _serviceEndpoint = result.Endpoint;
-        ClientEndpoint = result.ClientEndpoint;
-        ServerEndpoint = result.ServerEndpoint;
         EndpointType = type;
         Name = name;
 
         Endpoint = BuildEndpointString(_serviceEndpoint);
+        ClientEndpoint = result.ClientEndpoint ?? _serviceEndpoint;
+        ServerEndpoint = result.ServerEndpoint ?? _serviceEndpoint;
         AudienceBaseUrl = BuildAudienceBaseUrlEndWithSlash(_serviceEndpoint);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:132]: either:
```
public Uri ClientEndpoint
    {
        get => _clientEndpoint ?? _serviceEndpoint; 
    }
```

    and here: `_clientEndpoint = CheckScheme(result.ClientEndpoint);`

Or:
```

public Uri ClientEndpoint{get; init;}
```

    and here: `ClientEndpoint = CheckScheme(result.ClientEndpoint) ?? _serviceEndpoint;`
?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 7]
```diff
@@ -159,21 +158,20 @@ public ServiceEndpoint(Uri endpoint,
...
-        _serviceEndpoint = endpoint ?? throw new ArgumentNullException(nameof(endpoint));
-        CheckScheme(endpoint);
-
         _tokenCredential = credential ?? throw new ArgumentNullException(nameof(credential));
 
         EndpointType = endpointType;
         Name = name;
 
+        _serviceEndpoint = CheckScheme(endpoint);
+        _serverEndpoint = CheckScheme(serverEndpoint ?? endpoint);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs]: `serverEndpoint == null ? null : CheckScheme(serverEndpoint)`?

**Review by vicancy — APPROVED** [review]

**Pull request merged**